File: app/src/main/kotlin/tech/mmarca/openvitals/ui/theme/ActivityRecordingTheme.kt
package tech.mmarca.openvitals.ui.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.graphics.Color
import tech.mmarca.openvitals.domain.preferences.AppThemeMode
import tech.mmarca.openvitals.domain.preferences.isDarkTheme

val LocalActivityRecordingOutdoorMode = compositionLocalOf { false }

val LocalActivityRecordingOutdoorLightScheme = compositionLocalOf { false }

val RecordingOutdoorDarkColorScheme = darkColorScheme(
    primary = RecordingOutdoorAccent,
    onPrimary = Color(0xFF1A1A1A),
    primaryContainer = Color(0xFF3D2E00),
    onPrimaryContainer = RecordingOutdoorAccent,
    secondary = RecordingOutdoorAccentMuted,
    onSecondary = Color.Black,
    secondaryContainer = Color(0xFF3D2E00),
    onSecondaryContainer = RecordingOutdoorAccent,
    tertiary = RecordingOutdoorAccent,
    onTertiary = Color(0xFF1A1A1A),
    tertiaryContainer = Color(0xFF3D2E00),
    onTertiaryContainer = RecordingOutdoorAccent,
    error = Color(0xFFFFB4AB),
    errorContainer = Color(0xFF93000A),
    onError = Color(0xFF690005),
    onErrorContainer = Color(0xFFFFDAD6),
    background = RecordingOutdoorBackground,
    onBackground = Color.White,
    surface = RecordingOutdoorSurface,
    onSurface = Color.White,
    surfaceVariant = Color(0xFF1A1A1A),
    onSurfaceVariant = Color(0xFFBDBDBD),
    outline = Color(0xFF757575),
    outlineVariant = Color(0xFF424242),
    surfaceContainerLowest = RecordingOutdoorBackground,
    surfaceContainerLow = Color(0xFF0A0A0A),
    surfaceContainer = RecordingOutdoorSurface,
    surfaceContainerHigh = Color(0xFF1A1A1A),
    surfaceContainerHighest = Color(0xFF242424),
)

val RecordingOutdoorLightColorScheme = lightColorScheme(
    primary = RecordingOutdoorLightAccent,
    onPrimary = Color.White,
    primaryContainer = Color(0xFFFFDBC8),
    onPrimaryContainer = Color(0xFF341100),
    secondary = RecordingOutdoorLightAccent,
    onSecondary = Color.White,
    secondaryContainer = Color(0xFFFFDBC8),
    onSecondaryContainer = Color(0xFF341100),
    tertiary = RecordingOutdoorLightAccent,
    onTertiary = Color.White,
    tertiaryContainer = Color(0xFFFFDBC8),
    onTertiaryContainer = Color(0xFF341100),
    error = Color(0xFFBA1A1A),
    errorContainer = Color(0xFFFFDAD6),
    onError = Color.White,
    onErrorContainer = Color(0xFF410002),
    background = Color.White,
    onBackground = Color.Black,
    surface = Color.White,
    onSurface = Color.Black,
    surfaceVariant = Color.White,
    onSurfaceVariant = Color(0xFF424242),
    outline = Color(0xFF757575),
    outlineVariant = Color(0xFFBDBDBD),
    surfaceContainerLowest = Color.White,
    surfaceContainerLow = Color.White,
    surfaceContainer = Color.White,
    surfaceContainerHigh = Color.White,
    surfaceContainerHighest = Color(0xFFF5F5F5),
)

@Composable
fun ActivityRecordingTheme(
    outdoorModeEnabled: Boolean,
    appThemeMode: AppThemeMode,
    content: @Composable () -> Unit,
) {
    val outdoorUsesLightScheme = outdoorModeEnabled &&
        !appThemeMode.isDarkTheme(isSystemInDarkTheme())
    CompositionLocalProvider(
        LocalActivityRecordingOutdoorMode provides outdoorModeEnabled,
        LocalActivityRecordingOutdoorLightScheme provides outdoorUsesLightScheme,
    ) {
        if (outdoorModeEnabled) {
            MaterialTheme(
                colorScheme = if (outdoorUsesLightScheme) {
                    RecordingOutdoorLightColorScheme
                } else {
                    RecordingOutdoorDarkColorScheme
                },
                typography = AppTypography,
                shapes = MaterialTheme.shapes,
                content = content,
            )
        } else {
            content()
        }
    }
}

@Composable
fun activityRecordingAccentColor(): Color =
    when {
        !LocalActivityRecordingOutdoorMode.current -> WorkoutColor
        LocalActivityRecordingOutdoorLightScheme.current -> RecordingOutdoorLightAccent
        else -> RecordingOutdoorAccent
    }

@Composable
fun recordingOutdoorAccentForAppTheme(appThemeMode: AppThemeMode): Color =
    if (appThemeMode.isDarkTheme(isSystemInDarkTheme())) {
        RecordingOutdoorAccent
    } else {
        RecordingOutdoorLightAccent
    }

File: lib/ui/theme/activity_recording_theme.dart
import 'package:flutter/material.dart';

import '../../domain/preferences/app_theme_mode.dart';
import 'app_colors.dart';
import 'app_theme.dart';

/// Port of the Kotlin `ui/theme/ActivityRecordingTheme.kt`: a high-contrast
/// theme for recording outdoors in direct sunlight — amber-on-black when the
/// app is dark, burnt-orange-on-white when it is light.

/// Kotlin `RecordingOutdoorDarkColorScheme`.
const ColorScheme recordingOutdoorDarkColorScheme = ColorScheme(
  brightness: Brightness.dark,
  primary: AppColors.recordingOutdoorAccent,
  onPrimary: Color(0xFF1A1A1A),
  primaryContainer: Color(0xFF3D2E00),
  onPrimaryContainer: AppColors.recordingOutdoorAccent,
  secondary: AppColors.recordingOutdoorAccentMuted,
  onSecondary: Colors.black,
  secondaryContainer: Color(0xFF3D2E00),
  onSecondaryContainer: AppColors.recordingOutdoorAccent,
  tertiary: AppColors.recordingOutdoorAccent,
  onTertiary: Color(0xFF1A1A1A),
  tertiaryContainer: Color(0xFF3D2E00),
  onTertiaryContainer: AppColors.recordingOutdoorAccent,
  error: Color(0xFFFFB4AB),
  onError: Color(0xFF690005),
  errorContainer: Color(0xFF93000A),
  onErrorContainer: Color(0xFFFFDAD6),
  surface: AppColors.recordingOutdoorSurface,
  onSurface: Colors.white,
  onSurfaceVariant: Color(0xFFBDBDBD),
  outline: Color(0xFF757575),
  outlineVariant: Color(0xFF424242),
  surfaceContainerLowest: AppColors.recordingOutdoorBackground,
  surfaceContainerLow: Color(0xFF0A0A0A),
  surfaceContainer: AppColors.recordingOutdoorSurface,
  surfaceContainerHigh: Color(0xFF1A1A1A),
  surfaceContainerHighest: Color(0xFF242424),
);

/// Kotlin `RecordingOutdoorLightColorScheme`.
const ColorScheme recordingOutdoorLightColorScheme = ColorScheme(
  brightness: Brightness.light,
  primary: AppColors.recordingOutdoorLightAccent,
  onPrimary: Colors.white,
  primaryContainer: Color(0xFFFFDBC8),
  onPrimaryContainer: Color(0xFF341100),
  secondary: AppColors.recordingOutdoorLightAccent,
  onSecondary: Colors.white,
  secondaryContainer: Color(0xFFFFDBC8),
  onSecondaryContainer: Color(0xFF341100),
  tertiary: AppColors.recordingOutdoorLightAccent,
  onTertiary: Colors.white,
  tertiaryContainer: Color(0xFFFFDBC8),
  onTertiaryContainer: Color(0xFF341100),
  error: Color(0xFFBA1A1A),
  onError: Colors.white,
  errorContainer: Color(0xFFFFDAD6),
  onErrorContainer: Color(0xFF410002),
  surface: Colors.white,
  onSurface: Colors.black,
  onSurfaceVariant: Color(0xFF424242),
  outline: Color(0xFF757575),
  outlineVariant: Color(0xFFBDBDBD),
  surfaceContainerLowest: Colors.white,
  surfaceContainerLow: Colors.white,
  surfaceContainer: Colors.white,
  surfaceContainerHigh: Colors.white,
  surfaceContainerHighest: Color(0xFFF5F5F5),
);

/// Whether the outdoor theme should use its light (black-on-white) scheme.
/// Kotlin: `outdoorModeEnabled && !appThemeMode.isDarkTheme(isSystemInDarkTheme())`.
bool recordingOutdoorUsesLightScheme(
  BuildContext context, {
  required bool outdoorModeEnabled,
  required AppThemeMode appThemeMode,
}) {
  final systemInDarkTheme =
      MediaQuery.platformBrightnessOf(context) == Brightness.dark;
  return outdoorModeEnabled && !appThemeMode.isDarkTheme(systemInDarkTheme);
}

/// Kotlin `MaterialTheme.colorScheme.background` under the outdoor theme:
/// pure black for the dark scheme, pure white for the light one.
Color recordingOutdoorBackgroundColor({required bool outdoorUsesLightScheme}) =>
    outdoorUsesLightScheme ? Colors.white : AppColors.recordingOutdoorBackground;

/// Kotlin `recordingOutdoorAccentForAppTheme(appThemeMode)`: the accent the
/// outdoor toggle takes on while enabled, resolved against the app theme (not
/// the outdoor theme itself).
Color recordingOutdoorAccentForAppTheme(
  BuildContext context,
  AppThemeMode appThemeMode,
) {
  final systemInDarkTheme =
      MediaQuery.platformBrightnessOf(context) == Brightness.dark;
  return appThemeMode.isDarkTheme(systemInDarkTheme)
      ? AppColors.recordingOutdoorAccent
      : AppColors.recordingOutdoorLightAccent;
}

/// Kotlin `ActivityRecordingTheme`: swaps the ambient [Theme] for the
/// high-contrast outdoor scheme while outdoor mode is enabled; otherwise the
/// child renders under the app theme untouched.
class ActivityRecordingTheme extends StatelessWidget {
  const ActivityRecordingTheme({
    super.key,
    required this.outdoorModeEnabled,
    required this.appThemeMode,
    required this.child,
  });

  final bool outdoorModeEnabled;
  final AppThemeMode appThemeMode;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    if (!outdoorModeEnabled) return child;
    final usesLightScheme = recordingOutdoorUsesLightScheme(
      context,
      outdoorModeEnabled: outdoorModeEnabled,
      appThemeMode: appThemeMode,
    );
    // `AppTheme.themeFrom` keeps the app typography, as the Kotlin theme keeps
    // `AppTypography`.
    return Theme(
      data: AppTheme.themeFrom(
        usesLightScheme
            ? recordingOutdoorLightColorScheme
            : recordingOutdoorDarkColorScheme,
      ),
      child: child,
    );
  }
}
